rfcs/0109-nixpkgs-generated-code-policy.md

```diff
@@ -0,0 +1,210 @@
+
+# Future work
+[future]: #future-work
+
+## A possible 2nd step: Vendor generated code "out of tree"
+
+The first issue that remains after this RFC is generated code still bloats the Nixpkgs history.
+It would be nice to get it "out of tree" (outside the Nixpkgs repo) so this is no longer the case.
+In our shepherd discussions we had two ideas for how this might proceed.
+
+It was tempting to go straight to proposing one of these as part of the RFC proper,
+but they both contained enough hard-to-surmount issues that we figured it was better to start something more conservative first.
+
+### Alternative 1: Dump in other repo and fetch it
+
+We could opt to offload all generated code into a separate repository which would become an optional additional input to nixpkgs.
+This could be done via an extra `fetchTarball`, possibly a (somehow synced) channel or, in the presence of experimental features, a flake input.
+
+####  Drawbacks
+
+- This would be a truly breaking change to nixpkgs user interface:
+  Either an additional input would need to be provided or fetched (which wouldn't interact well with restrict-eval).
+
+- Generated code becomes a second class as the extra input would need to be optional for this reason.
+  This is problematic for central packages that use code generation already today (pandoc, cachix, …).
+
+- Similar Bootstrapping problems as the other alternative below: new generated code needs nixpkgs and a previous version of the generated code.
+
+- `builtins.fetch*` is a nuisance to deal with at the moment and would probably need to be improved to make this work.
+  E.g. gcrooting this evaluation only dependency could prove tricky without changes to Nix.
+
+- Extra bureaucracy would be involved with updating the generated repository and the reference to it in nixpkgs.
+  Additionally, special support in CI would be required for this.
+
+### Alternative 2: Nixpkgs itself becomes a derivation output
+
+This alternative implementation was proposed by @L-as at the meeting.
+The idea is that nixpkgs would become a derivation that builds a “regular” nixpkgs source tree by augmenting files available statically with code generation.
+
+The upside of this would be that there would only be one instance of IFD that can ever happen, namely when the source tree is built.
+The produced store path then would require no IFD, and it would be obvious what relates to IFD and what doesn't.
+
+In practice, IFD would not be necessary for users of nixpkgs if we can design a mechanism that allows the dynamically produced nixpkgs source tree to be used as a channel.
+Then the IFD would only need to be executed when working on nixpkgs.
+
+#### Drawbacks
+
+- This approach creates a bootstrapping problem for the entirety of nixpkgs, not just for the IFD parts.
+  It would be necessary to build the new nixpkgs source tree using an old version of the nixpkgs source tree.
```

Comment:
This sounds awful, I hate how manual `bootstrap-tools` are, doing it to the whole Nixpkgs feels even worse.

The question, though: Why can't only the generated parts of Nixpkgs be bootstrapped using the rest? That would better than either alternative, IMHO.
In fact, I wanted to do this for `all-packages.nix` in NixOS/nixpkgs#50643 (4 years before #140 and NixOS/nixpkgs#237439).

The ultimate idea (which was abandoned after the initial pushback) there was to eventually generate most of `all-packages.nix` by walking the Nixpkgs tree via a simple shell script.
So, `stdenv` and the knot of packages supporting it would have been left in a tiny hand-written equivalent of `all-packages.nix`, and the rest of `all-packages.nix` would have been auto-generated.

Reply:
> so the bootstrap problem already exists and can not be alleviated easily.

Well, yes, but there would be quite a difference between bootstrapping from an autogenerated `.nix` file that depends on current version of Nixpkgs (and which you can still read and understand, even if it is _technically_ programmatically generated) and bootstrapping from an autogenerated `.nix` file that depends on an outdated version of Nixpkgs that bootstraps from an autogenerated `.nix` file that depends on an even more outdated version of Nixpkgs, which bootstraps....

Reply:
That's why this is listed as an alternative to the actual proposal which does not require that!